Update of docker images for post-1.0 Veracruz

mathias-arm opened the pull request:
- Removed `sgx`, `tz` and `qemu` images (and related files)
- Update `veracruz/base` to Rust to 1.56.1
- Updated `RUN` statements to minimize layer sizes.
- Add `linux` targets (image is similar to `base`).
- `veracruz/base`, `veracruz/nitro` and `veracruz/linux` only involve `root` user but change permissions on directories and files so that an unpriviledged user can update (e.g. rustup or cargo) later.
- There is a `build-$(TEE)` target that will build `veracruz_image_$(TEE):$(USER)` image from `veracruz/$(TEE)` running with an unprivileged user. This works with `nitro` by creating a `docker` group so that the Docker control socket is accessible.
- Add `icecap/hacking` image
- Add `icecap` to CI
- Add `localci` image which works to build a `linux`, `nitro` and `icecap` (similar to `icecap/hacking`) which uses `$(USER)` permissions. All targets work for `x86_64`, and all but `icecap` work for `aarch64`.

mathias-arm committed:
Fix nitro image build

mathias-arm committed:
Document MacOS work-around

mathias-arm committed:
Remove files and targets related to TZ and SGX

mathias-arm committed:
Update "base" Dockerfiles and add linux for symmetry

mathias-arm committed:
Aligning Linux and Nitro build images

mathias-arm committed:
Add icecap hacking image

nspin commented:
https://github.com/veracruz-project/veracruz/pull/249 introduces the use of a Docker volume instead of a bind mount for the `/nix/store` in order to enable developing Veracruz on IceCap on MacOS.

We could incorporate those changes into this PR, or move them into a PR for this repository afterwards.

nspin commented:
Has the IceCap CI image creation been intentionally omitted from this PR?

mathias-arm commented:
> Has the IceCap CI image creation been intentionally omitted from this PR?

I am working on it

mathias-arm commented:
> [veracruz-project/veracruz#249](https://github.com/veracruz-project/veracruz/pull/249) introduces the use of a Docker volume instead of a bind mount for the `/nix/store` in order to enable developing Veracruz on IceCap on MacOS.

I forgot about this PR. I will try to add those changes.

mathias-arm committed:
Add icecap support to CI image

mathias-arm committed:
Add localci and use volume for icecap hacking

mathias-arm committed:
Size optimizations

mathias-arm committed:
New round of clean-ups and size optimization

nspin commented on icecap/ci/populate/default.nix:
```diff
@@ -0,0 +1,70 @@
+  roots = [
+    run
+    spec
+    pkgs.dev.icecap.rustc
+    pkgs.dev.icecap.cargo
```
These are unnecessary now that we don't use IceCap's rust toolchain. However, this is a minor note, because their presence doesn't really matter.

nspin commented on icecap/hacking/setup.sh:
```diff
@@ -0,0 +1,19 @@
+# See the `LICENSE_MIT.markdown` file in the Veracruz root directory for licensing
+# and copyright information.
+
+set -e
```
This could be changed to `set -eu` to make this script safer.

nspin commented on icecap/.dockerignore:
```diff
@@ -0,0 +1 @@
+/hacking/nix-root/
```
If we are using a volume for IceCap, this is no longer necessary.

dreemkiller commented on nitro/Dockerfile:
```diff
@@ -13,53 +13,60 @@
+        software-properties-common ; \
+    curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add - ; \
+    add-apt-repository "deb https://download.docker.com/linux/ubuntu bionic stable" ; \
+    apt-get update && apt-get install --no-install-recommends -y docker-ce-cli ; \
```
Should these two lines be combined with the above `apt-get update` so we only have to do one update?

mathias-arm replied:
If I remember correctly, the first `apt-get update` / `apt-get install` installs `add-apt-repository` that is used to add the docker repository...

dreemkiller commented on nitro/Dockerfile:
```diff
@@ -13,53 +13,60 @@
-    pip install awscli
+RUN pip install -U setuptools pip && \
+    pip install awscli && \
+    pip cache purge
```
Would it accomplish the same thing to use `--no-cache-dir` instead of purging afterward?

mathias-arm replied:
Probably. I did not know about `--no-cache-dir`

egrimley-arm committed:
Add "--init" to "docker run ... sleep inf".

This is to avoid a zombie apocalypse, though "ps aux" shows such
processes as "defunct" rather than "zombie".

To demonstrate the problem:

    docker run ... sleep inf
    docker exec -it ... /bin/bash
    ( ( sleep 1 ) & ) &
    ps auxw

To avoid the problem, replace the first line with one of:

    docker run ... /bin/sh -c 'sleep inf'

    docker run --init ... sleep inf

The latter option is used here.

mathias-arm committed:
Update based on PR feedback

mathias-arm committed:
Add "--init" to "docker run ... sleep inf" (continued)

Extend to other docker run calls

mathias-arm committed:
Update git for running container in GitHub Actions

mathias-arm committed:
Use 3 separate images for CI

The unified CI image was hitting some size limitations on GitHub Actions

mathias-arm commented:
@egrimley-arm was able to test the `localci` image with workspaces.
The new CI images were yielded a successful run with GitHub Actions for workspaces: https://github.com/mathias-arm/veracruz/actions/runs/1670121350
